src/sage/crypto/historical/enigma.py

```diff
@@ -0,0 +1,1148 @@
+    'HKVOWVPLLBC'
+    sage: one == two
+    True
+
+To decrypt an encrypted message we need to know the initial settings of the
+enigma used to encrypt the message. This can be done by just recreating that
+enigma::
+
+    sage: decrypt_enigma = Enigma.create([5, 3, 2], [5, 24, 11], [2, 4, 24], 2)
+    sage: decrypted = decrypt_enigma.en_de_crypt("HKVOWVPLLBC")
+    sage: decrypted
+    'GOODMORNING'
+
+If we encrypt multiple messages it may be stressful to recreate the exact state
+of the enigma to decrypt a specific message. In that case we can duplicate the
+enigma before we encrypt a message that we want to decrypt::
+
+    sage: enigma.en_de_crypt("MORE")
+    'QKYK'
+    sage: enigma.en_de_crypt("WORDS")
+    'VZGEV'
+    sage: decrypt_enigma = enigma.duplicate()
+    sage: encrypted = enigma.en_de_crypt("SECRET")
+    sage: encrypted
+    'NGZIFC'
+    sage: decrypted = decrypt_enigma.en_de_crypt(encrypted)
+    sage: decrypted
+    'SECRET'
+
+The enigma also supports a plugboard::
+
+    sage: p = Reflector([1, 3, 13, 19, 24, 2, 25, 8, 0, 17, 18, 20, 12, 4, 9,
+    ....: 11, 16, 21, 10, 6, 23, 15, 7, 22, 5, 14])
+    sage: enigma = Enigma.create([3, 4, 1], [12, 23, 1], [1, 8, 7], 1, p)
+    sage: decrypt_enigma = enigma.duplicate()
+    sage: encrypted = enigma.en_de_crypt("PLUGBOARD")
+    sage: encrypted
+    'YHKVADMPH'
+    sage: decrypted = decrypt_enigma.en_de_crypt("YHKVADMPH")
+    sage: decrypted
+    'PLUGBOARD'
+
+It may also be helpful to take a closer look at the representation of the used
+permutations. Each letter has a index beginning with 0 representing the letter
+A. If we use a plugboard [0, 1, 2, ...] we map every letter to itself (A -> A,
+B -> B, and so on)::
+
+    sage: p = Reflector([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
+    ....: 16, 17, 18, 19, 20, 21, 22, 23, 24, 25])
+    sage: e_wo_p = Enigma.create([3, 4, 1], [12, 23, 1], [1, 8, 7], 1)
+    sage: e_w_p = Enigma.create([3, 4, 1], [12, 23, 1], [1, 8, 7], 1, p)
+    sage: one = e_wo_p.en_de_crypt("PERMUTATION")
+    sage: one
+    'KQFCZJUITNF'
+    sage: two = e_w_p.en_de_crypt("PERMUTATION")
+    sage: two
+    'KQFCZJUITNF'
+    sage: one == two
+    True
+
+AUTHORS:
+
+- Nico Koeckerbauer and Benedek Kovacs (2024-06): initial version
+- Tobias Schneider (2024-06): proofreading and testing it via implementation of Turing-Bombe
+
+"""
+
+# ****************************************************************************
+#       Copyright (C) 2024 Nico Koeckerbauer <kolando@gmx.de>,
+#                           Tobias Schneider <bartv4der@gmail.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  https://www.gnu.org/licenses/
+# ****************************************************************************
+
+
+class Rotor:
```

Comment:
Sage classes should inherit from `SageObject`: https://doc.sagemath.org/html/en/developer/coding_in_python.html